File: react-native/library/src/starPlanet/components/TspCascader.js
import React, { useMemo, useState } from 'react';
import { Modal, Pressable, Text, View, StyleSheet, ScrollView } from 'react-native';
import { styles as shared, withTheme } from '../utils/shared';

function optId(opt, i) {
  return String(opt?.value ?? opt?.id ?? opt?.key ?? i);
}

function optLabel(opt) {
  return opt?.label ?? opt?.title ?? opt?.text ?? String(opt?.value ?? opt?.id ?? '');
}

function findPathLabels(options, valuePath) {
  const labels = [];
  let level = options;
  for (const v of valuePath) {
    const hit = (level || []).find((o, i) => optId(o, i) === String(v));
    if (!hit) break;
    labels.push(optLabel(hit));
    level = hit.children || [];
  }
  return labels;
}

/** Multi-level cascader — trigger + Modal multi-column panel. */
export function TspCascader({
  options = [],
  value = [],
  placeholder = '请选择',
  disabled = false,
  theme,
  onChange,
}) {
  const t = withTheme(theme);
  const opts = Array.isArray(options) ? options : [];
  const path = Array.isArray(value) ? value.map(String) : [];
  const [open, setOpen] = useState(false);
  const [draft, setDraft] = useState(path);

  const display = useMemo(() => {
    const labels = findPathLabels(opts, path);
    return labels.length ? labels.join(' / ') : '';
  }, [opts, path]);

  const columns = [];
  let level = opts;
  const walk = open ? draft : path;
  for (let i = 0; i <= walk.length; i += 1) {
    if (!Array.isArray(level) || level.length === 0) break;
    columns.push(level);
    const cur = walk[i];
    if (cur == null || cur === '') break;
    const hit = level.find((o, idx) => optId(o, idx) === String(cur));
    level = hit?.children || [];
  }

  const pick = (colIndex, opt, idx) => {
    const id = optId(opt, idx);
    const next = [...draft.slice(0, colIndex), id];
    setDraft(next);
    if (!opt.children?.length) {
      onChange?.(next, findPathLabels(opts, next));
      setOpen(false);
    }
  };

  return (
    <View style={[local.wrap, disabled && shared.disabled]}>
      <Pressable
        accessibilityRole="button"
        accessibilityLabel={display || placeholder}
        accessibilityState={{ disabled, expanded: open }}
        disabled={disabled}
        onPress={() => {
          if (disabled) return;
          setDraft(path);
          setOpen(true);
        }}
        style={[
          local.trigger,
          {
            backgroundColor: t.surfaceRaised,
            borderColor: t.borderDefault,
          },
        ]}
      >
        <Text
          style={[
            local.value,
            { color: display ? t.textPrimary : t.textTertiary },
            !display && { fontWeight: '600' },
          ]}
          numberOfLines={1}
        >
          {display || placeholder}
        </Text>
        <Text style={{ color: t.textTertiary }}>▾</Text>
      </Pressable>

      <Modal transparent animationType="fade" visible={open} onRequestClose={() => setOpen(false)}>
        <View style={local.maskRoot}>
          <Pressable
            accessibilityLabel="Close"
            accessibilityRole="button"
            style={local.maskFill}
            onPress={() => setOpen(false)}
          />
          <View
            accessibilityRole="list"
            accessibilityLabel="Cascader"
            style={[
              local.panel,
              { backgroundColor: t.surfaceRaised, borderColor: t.borderDefault },
            ]}
          >
            <View style={local.panelHeader}>
              <Text style={[shared.strong, { color: t.textPrimary, flex: 1 }]}>请选择</Text>
              <Pressable
                accessibilityRole="button"
                accessibilityLabel="Close"
                onPress={() => setOpen(false)}
                hitSlop={8}
                style={local.closeBtn}
              >
                <Text style={{ color: t.textSecondary, fontSize: 22, fontWeight: '700' }}>×</Text>
              </Pressable>
            </View>
            <ScrollView horizontal showsHorizontalScrollIndicator={false}>
              <View style={local.cols}>
                {columns.map((col, ci) => (
                  <ScrollView
                    key={ci}
                    style={[local.col, { borderRightColor: t.borderDefault }]}
                    showsVerticalScrollIndicator={false}
                  >
                    {col.map((opt, oi) => {
                      const id = optId(opt, oi);
                      const active = draft[ci] === id;
                      return (
                        <Pressable
                          key={id}
                          accessibilityRole="button"
                          accessibilityState={{ selected: active }}
                          onPress={() => pick(ci, opt, oi)}
                          style={[
                            local.option,
                            active && { backgroundColor: t.brandSubtle },
                          ]}
                        >
                          <Text
                            style={{
                              color: active ? t.brandPrimary : t.textPrimary,
                              fontWeight: '700',
                            }}
                            numberOfLines={1}
                          >
                            {optLabel(opt)}
                          </Text>
                        </Pressable>
                      );
                    })}
                  </ScrollView>
                ))}
              </View>
            </ScrollView>
          </View>
        </View>
      </Modal>
    </View>
  );
}

const local = StyleSheet.create({
  wrap: { width: '100%' },
  trigger: {
    minHeight: 48,
    borderWidth: 1,
    borderRadius: 16,
    paddingHorizontal: 14,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    gap: 8,
  },
  value: { flex: 1, fontSize: 15, fontWeight: '700' },
  maskRoot: { flex: 1, backgroundColor: 'rgba(23,58,98,0.32)', justifyContent: 'flex-end' },
  maskFill: { ...StyleSheet.absoluteFillObject },
  panel: {
    borderTopLeftRadius: 24,
    borderTopRightRadius: 24,
    borderWidth: 1,
    maxHeight: '62%',
    zIndex: 1,
  },
  panelHeader: {
    minHeight: 52,
    flexDirection: 'row',
    alignItems: 'center',
    paddingHorizontal: 16,
    gap: 8,
  },
  closeBtn: { width: 36, height: 36, alignItems: 'center', justifyContent: 'center' },
  cols: { flexDirection: 'row', minHeight: 240, paddingBottom: 12 },
  col: {
    width: 140,
    borderRightWidth: StyleSheet.hairlineWidth,
    maxHeight: 280,
  },
  option: {
    minHeight: 44,
    paddingHorizontal: 12,
    justifyContent: 'center',
    marginHorizontal: 6,
    borderRadius: 12,
  },
});

File: kuikly/library/shared/src/commonMain/kotlin/com/techskillplanet/phonics/controls/TspCascader.kt
package com.techskillplanet.phonics.controls

import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import com.tencent.kuikly.compose.foundation.background
import com.tencent.kuikly.compose.foundation.border
import com.tencent.kuikly.compose.foundation.clickable
import com.tencent.kuikly.compose.foundation.layout.Arrangement
import com.tencent.kuikly.compose.foundation.layout.Column
import com.tencent.kuikly.compose.foundation.layout.Row
import com.tencent.kuikly.compose.foundation.layout.fillMaxWidth
import com.tencent.kuikly.compose.foundation.layout.heightIn
import com.tencent.kuikly.compose.foundation.layout.padding
import com.tencent.kuikly.compose.foundation.layout.width
import com.tencent.kuikly.compose.foundation.shape.RoundedCornerShape
import com.tencent.kuikly.compose.material3.Text
import com.tencent.kuikly.compose.ui.Alignment
import com.tencent.kuikly.compose.ui.Modifier
import com.tencent.kuikly.compose.ui.graphics.Color
import com.tencent.kuikly.compose.ui.text.font.FontWeight
import com.tencent.kuikly.compose.ui.unit.dp
import com.tencent.kuikly.compose.ui.unit.sp
import com.techskillplanet.phonics.theme.PhonicsColors

data class TspCascaderOption(
    val value: String,
    val label: String,
    val children: List<TspCascaderOption> = emptyList(),
)

@Composable
fun TspCascader(
    theme: PhonicsColors,
    modifier: Modifier = Modifier,
    options: List<TspCascaderOption> = emptyList(),
    value: List<String> = emptyList(),
    placeholder: String = "请选择",
    disabled: Boolean = false,
    onChange: (List<String>, List<String>) -> Unit = { _, _ -> },
) {
    val open = remember { mutableStateOf(false) }
    val draft = remember { mutableStateOf(value) }
    val labels = findPathLabels(options, value)
    val display = if (labels.isEmpty()) placeholder else labels.joinToString(" / ")
    Column(modifier = modifier.fillMaxWidth()) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(min = 48.dp)
                .background(Color(theme.surfaceRaised), RoundedCornerShape(16.dp))
                .border(2.dp, Color(theme.borderDefault), RoundedCornerShape(16.dp))
                .clickable(enabled = !disabled) {
                    draft.value = value
                    open.value = !open.value
                }
                .padding(horizontal = 14.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            Text(
                text = display,
                color = Color(
                    when {
                        disabled -> theme.textTertiary
                        labels.isEmpty() -> theme.textTertiary
                        else -> theme.textPrimary
                    },
                ),
                fontWeight = FontWeight.Bold,
                fontSize = 15.sp,
                modifier = Modifier.weight(1f),
            )
            Text(text = "▾", color = Color(theme.textTertiary), fontWeight = FontWeight.Bold)
        }
        if (open.value && !disabled) {
            val columns = buildColumns(options, draft.value)
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 8.dp)
                    .background(Color(theme.surfaceSubtle), RoundedCornerShape(14.dp))
                    .border(1.dp, Color(theme.borderDefault), RoundedCornerShape(14.dp))
                    .padding(8.dp),
            ) {
                columns.forEachIndexed { colIndex, col ->
                    Column(
                        modifier = Modifier
                            .width(120.dp)
                            .padding(end = 6.dp),
                    ) {
                        col.forEach { opt ->
                            val active = draft.value.getOrNull(colIndex) == opt.value
                            Text(
                                text = opt.label,
                                color = Color(if (active) theme.brandPrimary else theme.textPrimary),
                                fontWeight = FontWeight.Bold,
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .background(
                                        Color(if (active) theme.brandSubtle else 0x00000000),
                                        RoundedCornerShape(10.dp),
                                    )
                                    .clickable {
                                        val next = draft.value.take(colIndex) + opt.value
                                        draft.value = next
                                        if (opt.children.isEmpty()) {
                                            open.value = false
                                            onChange(next, findPathLabels(options, next))
                                        }
                                    }
                                    .padding(horizontal = 10.dp, vertical = 10.dp),
                            )
                        }
                    }
                }
            }
        }
    }
}

private fun buildColumns(
    options: List<TspCascaderOption>,
    walk: List<String>,
): List<List<TspCascaderOption>> {
    val columns = mutableListOf<List<TspCascaderOption>>()
    var level: List<TspCascaderOption>? = options
    var i = 0
    while (level != null && level.isNotEmpty()) {
        columns.add(level)
        if (i >= walk.size) break
        val cur = walk[i]
        val hit = level.find { it.value == cur }
        level = hit?.children
        i += 1
    }
    return columns
}

private fun findPathLabels(options: List<TspCascaderOption>, path: List<String>): List<String> {
    val labels = mutableListOf<String>()
    var level = options
    for (v in path) {
        val hit = level.find { it.value == v } ?: break
        labels.add(hit.label)
        level = hit.children
    }
    return labels
}
